geomstats/geometry/pre_shape.py

```diff
@@ -0,0 +1,405 @@
+        tangent_vec : array-like, shape=[..., m, k]
+            Tangent vector to the pre-shape space at `base_point`.
+        base_point : array-like, shape=[..., m, k]
+            Point on the pre-shape space.
+
+        Returns
+        -------
+        vertical : array-like, shape=[..., m, k]
+            Vertical component of `tangent_vec`.
+        """
+        transposed_point = Matrices.transpose(base_point)
+        left_term = gs.matmul(base_point, transposed_point)
+        right_term = gs.matmul(tangent_vec, transposed_point) - gs.matmul(
+            base_point, Matrices.transpose(tangent_vec))
+        skew = gs.linalg.solve_sylvester(left_term, left_term, right_term)
+
+        return gs.matmul(skew, base_point)
+
+    @classmethod
+    def horizontal_projection(cls, tangent_vec, base_point):
+        r"""Project to horizontal subspace.
+
+        Compute the horizontal component of a tangent vector :math: `w` at a
+        base point :math: `x` by removing the vertical component.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., m, k]
+            Tangent vector to the pre-shape space at `base_point`.
+        base_point : array-like, shape=[..., m, k]
+            Point on the pre-shape space.
+
+        Returns
+        -------
+        horizontal : array-like, shape=[..., m, k]
+            Horizontal component of `tangent_vec`.
+        """
+        return tangent_vec - cls.vertical_projection(tangent_vec, base_point)
+
+    def is_horizontal(self, tangent_vec, base_point, atol=TOLERANCE):
+        """Check whether the tangent vector is horizontal at base_point.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., m, k]
+            Tangent vector.
+        base_point : array-like, shape=[..., m, k]
+            Point on the manifold.
+            Optional, default: none.
+        atol : float
+            Absolute tolerance.
+            Optional, default: 1e-6.
+
+        Returns
+        -------
+        is_tangent : bool
+            Boolean denoting if tangent vector is horizontal.
+        """
+        product = gs.matmul(tangent_vec, Matrices.transpose(base_point))
+        is_tangent = self.is_tangent(tangent_vec, base_point, atol)
+        is_symmetric = Matrices.is_symmetric(product, atol)
+        return gs.logical_and(is_tangent, is_symmetric)
+
+    def realign(self, point, base_point):
+        """Realign point to base_point.
+
+        Find the optimal rotation R in SO(m) such that the base point and
+        R.point are well positioned.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., m, k]
+            Point on the manifold.
+        base_point : array-like, shape=[..., m, k]
+            Point on the manifold.
+
+        Returns
+        -------
+        aligned : array-like, shape=[..., m, k]
+            R.point.
+        """
+        M = gs.matmul(point, Matrices.transpose(base_point))
```

Comment:
Let's try to avoid uppercase variable names when possible: uppercase letters are reserved to constant variables in python.
Maybe:
M -> mat
U, S, Vh -> mat_u, mat_s, mat_vh

https://www.python.org/dev/peps/pep-0008/#constants

Reply:
(single letter variables are also a pain when we try to do global renaming, as trying to rename all the "M"'s automatically is not always possible! - can do it with vscode, but not everybody uses it).